File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/businessinbox/BusinessInboxDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.businessinbox

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Wire shapes for the business-side inbox:
 *  - `GET /api/chat/business/:businessUserId/rooms`
 *    (`backend/routes/chats.js:756` builds the row);
 *  - `GET /api/businesses/:businessId/matched-posts`
 *    (`backend/routes/businesses.js:4380` selects the columns).
 *
 * Mirrors iOS `BusinessInboxDTOs.swift`.
 */
@JsonClass(generateAdapter = true)
data class BusinessInboxRoomsResponse(
    val rooms: List<BusinessInboxRoomDto> = emptyList(),
    val total: Int? = null,
    val totalUnread: Int? = null,
)

/**
 * One room the business identity participates in. The serializer flattens
 * the counterpart onto `other_participant_*` fields rather than nesting a
 * participants array.
 */
@JsonClass(generateAdapter = true)
data class BusinessInboxRoomDto(
    val id: String,
    @Json(name = "room_type") val roomType: String? = null,
    @Json(name = "room_name") val roomName: String? = null,
    @Json(name = "last_message_at") val lastMessageAt: String? = null,
    @Json(name = "last_message_preview") val lastMessagePreview: String? = null,
    @Json(name = "unread_count") val unreadCount: Int? = null,
    @Json(name = "other_participant_name") val otherParticipantName: String? = null,
    @Json(name = "other_participant_username") val otherParticipantUsername: String? = null,
)

@JsonClass(generateAdapter = true)
data class BusinessMatchedPostsResponse(
    val posts: List<BusinessMatchedPostDto> = emptyList(),
    val pagination: BusinessMatchedPostsPageDto? = null,
)

@JsonClass(generateAdapter = true)
data class BusinessMatchedPostsPageDto(
    val page: Int? = null,
    @Json(name = "page_size") val pageSize: Int? = null,
    @Json(name = "total_count") val totalCount: Int? = null,
    @Json(name = "total_pages") val totalPages: Int? = null,
)

/** One neighborhood post whose `matched_business_ids` contains this business. */
@JsonClass(generateAdapter = true)
data class BusinessMatchedPostDto(
    val id: String,
    val title: String? = null,
    val content: String? = null,
    @Json(name = "post_type") val postType: String? = null,
    @Json(name = "like_count") val likeCount: Int? = null,
    @Json(name = "comment_count") val commentCount: Int? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    val creator: BusinessMatchedPostCreatorDto? = null,
)

@JsonClass(generateAdapter = true)
data class BusinessMatchedPostCreatorDto(
    val id: String? = null,
    val name: String? = null,
    val username: String? = null,
    @Json(name = "profile_picture_url") val profilePictureUrl: String? = null,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/BusinessInbox/BusinessInboxDTOs.swift
//
//  BusinessInboxDTOs.swift
//  Pantopus
//
//  Wire shapes for the business-side inbox:
//    · `GET /api/chat/business/:businessUserId/rooms`
//      (`backend/routes/chats.js:756-775` builds the row);
//    · `GET /api/businesses/:businessId/matched-posts`
//      (`backend/routes/businesses.js:4380` selects the columns).
//
//  Snake_case is mapped explicitly — `APIClient` does not apply
//  `convertFromSnakeCase`.
//

import Foundation

// MARK: - Messages

/// Envelope for the business chat inbox.
public struct BusinessInboxRoomsResponse: Decodable, Sendable {
    public let rooms: [BusinessInboxRoomDTO]
    public let total: Int?
    public let totalUnread: Int?

    enum CodingKeys: String, CodingKey {
        case rooms, total, totalUnread
    }
}

/// One room the business identity participates in. The serializer flattens
/// the counterpart onto `other_participant_*` fields rather than nesting a
/// participants array.
public struct BusinessInboxRoomDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let roomType: String?
    public let roomName: String?
    public let lastMessageAt: String?
    public let lastMessagePreview: String?
    public let unreadCount: Int?
    public let otherParticipantName: String?
    public let otherParticipantUsername: String?

    enum CodingKeys: String, CodingKey {
        case id
        case roomType = "room_type"
        case roomName = "room_name"
        case lastMessageAt = "last_message_at"
        case lastMessagePreview = "last_message_preview"
        case unreadCount = "unread_count"
        case otherParticipantName = "other_participant_name"
        case otherParticipantUsername = "other_participant_username"
    }
}

// MARK: - Mentions

/// Envelope for the matched-posts ("Mentions") section.
public struct BusinessMatchedPostsResponse: Decodable, Sendable {
    public let posts: [BusinessMatchedPostDTO]
    public let pagination: BusinessMatchedPostsPageDTO?
}

public struct BusinessMatchedPostsPageDTO: Decodable, Sendable, Hashable {
    public let page: Int?
    public let pageSize: Int?
    public let totalCount: Int?
    public let totalPages: Int?

    enum CodingKeys: String, CodingKey {
        case page
        case pageSize = "page_size"
        case totalCount = "total_count"
        case totalPages = "total_pages"
    }
}

/// One neighborhood post whose `matched_business_ids` contains this
/// business.
public struct BusinessMatchedPostDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let title: String?
    public let content: String?
    public let postType: String?
    public let likeCount: Int?
    public let commentCount: Int?
    public let createdAt: String?
    public let creator: BusinessMatchedPostCreatorDTO?

    enum CodingKeys: String, CodingKey {
        case id, title, content, creator
        case postType = "post_type"
        case likeCount = "like_count"
        case commentCount = "comment_count"
        case createdAt = "created_at"
    }
}

public struct BusinessMatchedPostCreatorDTO: Decodable, Sendable, Hashable {
    public let id: String?
    public let name: String?
    public let username: String?
    public let profilePictureUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, name, username
        case profilePictureUrl = "profile_picture_url"
    }
}
